Add col names prop to Table & add col names to UserShortLinksSection

Coteh opened the pull request:
Fixes #702.

## Current Behavior ( Optional for new feature )
### Description
Column widths in "Favorites" section is not consistent. If a *long* long link is shown on a cell within the left column, it will stretch the whole column out for that page.
### Screenshots
See PR #702.

## New Behavior
### Description
Allocated 70% of the width to the long link column and 30% of the width to the alias column (only one of the CSS rules are needed but I added both to be explicit). This will fix the problem.
### Screenshots
![result](https://user-images.githubusercontent.com/3276350/80551302-08419380-8991-11ea-8328-ac7ab9ef2a74.gif)

## TODO
- [x] Write new test(s) for `Table`
- [x] ~Write new test(s) for `UserShortLinksSection`~ decided not to write tests here as the only major change to this component was style-related

byliuyang commented on frontend/src/component/pages/shared/UserShortLinksSection.scss:
```diff
@@ -14,4 +14,12 @@
+    width: 70%;
+  }
+  
+  .table-cell.alias {
+    width: 30%;
+  }
```
Let's remove this for now.

byliuyang commented on frontend/src/component/ui/Table.tsx:
```diff
@@ -1,22 +1,24 @@
 interface IProps {
   headers?: ReactChild[];
   rows?: ReactChild[][];
+  colNames?: string[];
```
```suggestion
  colClassNames?: string[];
```

Coteh replied:
My next commit will contain this fix.

byliuyang commented on frontend/src/component/ui/Table.tsx:
```diff
@@ -1,22 +1,24 @@
         {headers.map((cell: ReactChild, cellIndex: number) => {
           return (
-            <th key={`cell-${cellIndex}`} className="table-cell">
+            <th key={`cell-${cellIndex}`} className={!colNames ? "table-cell" : classNames("table-cell", colNames[cellIndex])}>
```
Let's simplify this one. We want to remove table-cell and keep column class Names only.

rohithbalaji123 commented:
@Coteh , can you please check the case for long alias as well?

![alias_bug](https://user-images.githubusercontent.com/18581705/80581597-02988c00-8a2b-11ea-9947-53e834fcfb66.gif)

Coteh commented:
Now accommodate for long aliases.

![result](https://user-images.githubusercontent.com/3276350/80648808-d424ac00-8a3e-11ea-8714-accc57da2f8a.gif)

Coteh commented on frontend/src/component/ui/Table.tsx:
```diff
@@ -1,22 +1,24 @@
 import React, { Component, ReactChild } from 'react';
 
 import './Table.scss';
+import classNames from 'classnames';
```
Remove this import

Coteh replied:
Done

byliuyang commented on frontend/src/component/pages/shared/UserShortLinksSection.scss:
```diff
@@ -11,7 +11,13 @@
     word-break: break-all;
   }
+
+  td.cell-long,
+  th.cell-long {
```
```suggestion
  .cell-long {
```

Coteh replied:
Already done, but GitHub still thinks this is a change to be requested for some reason.

Coteh commented:
Spent a bit of time playing around with `table-layout: auto` and `table-layout: fixed`. Although both options seemed to be unaffected by the table content after the `word-break: all` property is set, I ultimately went with `table-layout: fixed` as [it guarantees that the table layout will be unaffected by table contents](https://css-tricks.com/fixing-tables-long-strings/). Also see [this](https://css-tricks.com/almanac/properties/t/table-layout/) too: _"The column values used are based on widths defined on columns or cells for the first row of the table."_

Additionally, I added some tests to `Table` for the col class names. However, I couldn't find a way to test column width using `react-testing-library` (column sizes appear to be 0 when rendered using the test framework), and I believe this is by design as sizes of elements rendered is more of a browser implementation detail and I don't believe this testing library is designed for testing implementation details in this manner. So therefore I just have the two tests in this PR.

Ready for review.

Coteh commented:
I noticed that `prettier` didn’t seem to like my TypeScript conditional operator. However, after upgrading `prettier` to the latest version (1.19.1) it went away
```
src/component/ui/Table.tsx
[error] src/component/ui/Table.tsx: SyntaxError: Expression expected. (20:68)
[error]   18 |         {headers.map((cell: ReactChild, cellIndex: number) => {
[error]   19 |           return (
[error] > 20 |             <th key={`cell-${cellIndex}`} className={colClassNames?.[cellIndex]}>
[error]      |                                                                    ^
[error]   21 |               {cell}
[error]   22 |             </th>
[error]   23 |           );
```
it looked like this

@byliuyang Should we proceed with upgrading `prettier` in this PR?

Coteh commented:
Ready for review

byliuyang commented on frontend/src/component/ui/Table.spec.tsx:
```diff
@@ -129,4 +129,60 @@ describe('Table component', () => {
+    const tableRows = container.querySelectorAll('tbody > tr');
+    for (let row = 0; row < sampleRows.length; row++) {
+      const curRowData = tableRows[row].querySelectorAll('td');
+      for (let column = 0; column < sampleRows[row].length; column++) {
```
```suggestion

      for (let column = 0; column < sampleRows[row].length; column++) {
```

byliuyang commented on frontend/src/component/ui/Table.spec.tsx:
```diff
@@ -129,4 +129,60 @@ describe('Table component', () => {
+    const tableRows = container.querySelectorAll('tbody > tr');
+    for (let row = 0; row < sampleRows.length; row++) {
+      const curRowData = tableRows[row].querySelectorAll('td');
+      for (let column = 0; column < sampleRows[row].length; column++) {
```
```suggestion

      for (let column = 0; column < sampleRows[row].length; column++) {
```

byliuyang commented on frontend/src/component/ui/Table.tsx:
```diff
@@ -25,33 +32,43 @@ export class Table extends Component<IProps> {
     return row.map((cell: ReactChild, cellIndex: number) => {
       return (
-        <td key={`cell-${cellIndex}`} className="table-cell">
+        <td key={`cell-${cellIndex}`} className={colClassNames?.[cellIndex]}>
```
Please use 
```typescript
classNames({
          colClassNames?.[cellIndex]: true
        })
```
So that we can extend the style in the future.

byliuyang commented on frontend/src/component/pages/shared/UserShortLinksSection.tsx:
```diff
@@ -38,6 +38,7 @@ export class UserShortLinksSection extends Component<IProps> {
           <Table
             headers={['Long Link', 'Alias']}
             rows={this.createTableRows()}
+            colClassNames={['cell-long', 'cell-alias']}
```
To recommend passing in `columnWidths` instead of `colClassNames` because `colClassNames` is implementation detail of table component. The component using relays on those classes may break when we change the internal of the table.

Coteh replied:
The `colClassNames` property was added to `Table` component so that columns can have their widths styled in via SCSS, but if we're going to pass the widths directly instead, then `colClassNames` should be deleted and instead replaced with `columnWidths`, which will add the widths directly to the columns.

byliuyang replied:
@Coteh Please remove `colClassNames` to enforce encapsulation. Thank you!

Coteh commented:
Changed to widths. Should appear the same as the previous implementation, just now the widths are set explicitly to each element.

![result](https://user-images.githubusercontent.com/3276350/82703256-8c5e0280-9c41-11ea-8b2c-1ee8f60a845b.gif)
